c/src/ml-api-common.c

```diff
@@ -1158,3 +1158,125 @@ ml_strerror (int errnum)
     return NULL;
   return strerrors[errnum];
 }
+
+/**
+ * @brief Creates an option and returns the instance a handle.
+ */
+int
+ml_option_create (ml_option_h * option)
+{
+  ml_option_s *_option;
+
+  check_feature_state (ML_FEATURE);
+
+  if (!option) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'option' is NULL. It should be a valid ml_option_h");
+  }
+
+  _option = g_new0 (ml_option_s, 1);
+  if (_option == NULL)
+    _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
+        "Failed to allocate memory for the option handle. Out of memory?");
+
+  _option->option_table =
+      g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  if (_option->option_table == NULL) {
+    _ml_error_report
+        ("Failed to create a new table for ml_option. Out of memory?");
+    g_free (_option);
+    return ML_ERROR_OUT_OF_MEMORY;
+  }
+
+  *option = (ml_option_h *) _option;
+  return ML_ERROR_NONE;
+}
+
+/**
+ * @brief Frees the given handle of a ml_option.
+ */
+int
+ml_option_destroy (ml_option_h option)
+{
+  ml_option_s *_option;
+  ml_option_value_s *_option_value;
+  GHashTableIter iter;
+  gchar *key;
+
+  check_feature_state (ML_FEATURE);
+
+  if (!option) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'option' is NULL. It should be a valid ml_option_h, which should be created by ml_option_create().");
+  }
+
+  _option = (ml_option_s *) option;
+
+  /** for each key, destroy its value */
+  g_hash_table_iter_init (&iter, _option->option_table);
+  while (g_hash_table_iter_next (&iter, (gpointer *) & key,
+          (gpointer *) & _option_value)) {
+    if (_option_value->destroy) {
```

Comment:
Let's add internal function to release data in hashtable and remove these iterations.
- `g_hash_table_destroy (_option->option_table);` will call this.
- line 1268 `g_hash_table_remove (_option->option_table, key);` also call this.
-  
```
static void
_ml_option_value_free (void *data)
{
  ml_option_value_s *_value;

  _value = (ml_option_value_s *) data;
  if (_value) {
    if (_value->destroy)
      _value->destroy (_value->value);
    g_free (_value);
  }
}
```

Reply:
Thanks, this makes code way more clean.